var/spack/repos/builtin/packages/netcdf-cxx4/package.py

```diff
@@ -22,8 +26,13 @@ class NetcdfCxx4(AutotoolsPackage):
 
     force_autoreconf = True
 
-    def configure_args(self):
-        return ['CPPFLAGS=-I' + self.spec['netcdf'].prefix.include]
+    def flag_handler(self, name, flags):
+        if name in ['cflags', 'fflags'] and '+pic' in self.spec:
+            flags.append(self.compiler.pic_flag)
+        elif name == 'cppflags':
+            flags.append('-I' + self.spec['netcdf'].prefix.include)
```

Comment:
@bebosudo @skosukhin 

https://github.com/spack/spack/pull/9647 was just proposed. It looks like it could fold into this PR. One of these will conflict with the other. Would you be able to do that?

Reply:
I can update my PR when/if this is merged. I can't help but say that I am very disappointed that this `+pic` thing infects netcdf packages too (yes, it was there before but at least it was disabled by default). Could anybody explain to me why this variant is so useful? Is it all about having static libraries with PIC? Because for me this is much more a confusing this than a useful one:
1. If a package has a variant `a` and I install a package with `~a` and `+a`, I expect to get two installations with different properties. Is it really the case for the variant `pic` most of the time? Don't I still get PIC when I set `~pic` (having in mind that usually the building systems manage to detect and to set PIC flags on their own)?
2. PIC flags for Fortran and C compilers are not always the same, so we can't even build the package by default. Can we at least have `~pic` by default?

BTW, does `netcdf-cxx` use Fortran compiler? If no, why does its `flag_handler` modifies `fflags`?

Reply:
> Could anybody explain to me why this variant is so useful? Is it all about having static libraries with PIC?

That's the goal

> Don't I still get PIC when I set ~pic (having in mind that usually the building systems manage to detect and to set PIC flags on their own)?

That's a fair point. It likely differ between different packages and how it is handled in their `configure` scripts. It's possible that a package will add `PIC` without this. I'm assuming @bebosudo did this because it didn't happen automatically for them.

It's also true that a sensible interpretation of `~pic` would be that the static libraries are guaranteed to be compiled without `PIC`. To actually support that, a configuration option would likely be required.

It is worth us having a conversation about how to establish conventions for `shared`, `static`, and `pic` variants. Part of the issue is that I'd like to combine conventions like "`~shared` means you don't build shared libraries and `~pic` means that all static libs must not be PIC" with convenience functionality in build system classes.

Actually, there is already at least one conversation on this at https://github.com/spack/spack/issues/5269, which I need to refresh on. If there are others, (that aren't linked from there) that you think are relevant, could you let me know?

I don't want to tie this PR to the resolution of that conversation, so I'm inclined to push ahead with it and address these sorts of issues later: there are many packages which currently have this behavior, and I don't think adding this will create significant extra work. Does that sound reasonable?

Reply:
@skosukhin 
> I can't help but say that I am very disappointed that this `+pic` thing infects netcdf packages too (yes, it was there before but at least it was disabled by default).

I wrote this PR because I was facing this kind of error: Unidata/netcdf-fortran/issues/35

> 1. If a package has a variant `a` and I install a package with `~a` and `+a`, I expect to get two installations with different properties. Is it really the case for the variant `pic` most of the time? Don't I still get PIC when I set `~pic` (having in mind that usually the building systems manage to detect and to set PIC flags on their own)?

What do you propose to fix this?

> 2. PIC flags for Fortran and C compilers are not always the same, so we can't even build the package by default. Can we at least have `~pic` by default?

I'm sorry, I don't get what's the problem. `-fPIC` is the same option on both C and Fortran compiler, and I take it with: `self.compiler.pic_flag`.

> BTW, does `netcdf-cxx` use Fortran compiler? If no, why does its `flag_handler` modifies `fflags`?

You're right, I'll remove it.


@scheibelp 
I get the problem. What do you think is better at the moment?

Reply:
@bebosudo 

1. I don't really understand how this PR solves the referenced problem. As far as I understand, you're trying to build `netcdf-fortran` using a static, no-PIC version of `netcdf` from `/usr/local/lib`. It's hardly that that `netcdf` is installed with Spack because if it was there would be either a shared version of the library (`spack install netcdf+shared`), which would be preferred over the static one by the linker, or only a static version but compiled with PIC (`spack install netcdf~shared`) because Spack already injects the PIC flag when `~shared`. So, it's a question to the providers of the `netcdf` installation on your system.

2. I suggest that we always have `+shared` by default and always inject PIC flags only when `~shared` (and not have `pic` variant at all). Although I don't really understand why people would like to build only static versions (but that's my problem :).

3. Sometimes we need to set up [mixed toolchains](https://spack.readthedocs.io/en/latest/getting_started.html?#mixed-toolchains). For example, [NAG Fortran compiler usually works together with `gcc` as the C compiler](https://spack.readthedocs.io/en/latest/getting_started.html?#nag):  `...FC=nagfor CC=gcc...`. In this case, `self.compiler.pic_flag` returns `-PIC`:
https://github.com/spack/spack/blob/930d6af815b3b87de31d1d704c63e5e59dc5d73f/lib/spack/spack/compilers/nag.py#L40-L42
The problem is that `gcc` recieves `-PIC` too but it can't understand it and fails.

Reply:
1. The issue I linked above was only to share the kind of message I got: I was actually compiling an application (#9655) above netcdf-fortran, which raised that error because netcdf-fortran wasn't compiled with fPIC.

2. Ok.

3. You're right, this is a problem.
What do you suggest to get the proper pic flags for each compiler? Because I tried looking in `self.compiler.fc`, but it's a subclass of `str` and it doesn't have and `pic_flag`.
@scheibelp have you already done something like this in mixed toolchains?

Reply:
> Spack already injects the PIC flag when ~shared. 

@skosukhin I don't think that happens automatically (if it does, I don't see where), do you know where Spack does that?

Selecting the appropriate `pic_flag` for a mixed toolchain sounds like a new issue, I don't immediately have a good solution for it. 

IMO a good short-term solution would be per-package compiler flags, which is another PR somewhere, in which case we could at least set `pic` specifically for `netcdf` for now on systems which aren't using toolchains with incompatible PIC flags.

> I suggest that we always have `+shared` by default and always inject PIC flags only when `~shared`

If you are suggesting a convention, that one may be problematic when users want to compile static libs without PIC (more on that at https://github.com/spack/spack/issues/5269)

Reply:
> > Spack already injects the PIC flag when ~shared.
>
> I don't think that happens automatically (if it does, I don't see where), do you know where Spack does that?

I think he's referring to these lines: https://github.com/spack/spack/blob/6374e0b7a70e74d9d8c385ce20855c3d64a68111/var/spack/repos/builtin/packages/netcdf/package.py#L160-L163

Regarding this PR, should we keep both the `shared` and `pic` variants in the `netcdf` package?
Is there something else to complete?

Reply:
@scheibelp  Any update? What is missing to close this PR?

Reply:
@bebosudo this looks good. I'll merge it after SC: 11/14 or 11/15. Apologies for the delay.